pkg/reconciler/workspace/controller.go

```diff
@@ -0,0 +1,385 @@
+const targetShardName = "targetShard"
+const unschedulable = "unschedulable"
+
+func NewController(
+	kcpClient kcpclient.ClusterInterface,
+	workspaceInformer tenancyinformer.WorkspaceInformer,
+	workspaceShardInformer tenancyinformer.WorkspaceShardInformer,
+) (*Controller, error) {
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+
+	c := &Controller{
+		queue:                 queue,
+		kcpClient:             kcpClient,
+		workspaceIndexer:      workspaceInformer.Informer().GetIndexer(),
+		workspaceLister:       workspaceInformer.Lister(),
+		workspaceShardIndexer: workspaceShardInformer.Informer().GetIndexer(),
+		workspaceShardLister:  workspaceShardInformer.Lister(),
+	}
+
+	workspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueue(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
+	})
+	if err := c.workspaceIndexer.AddIndexers(map[string]cache.IndexFunc{
+		targetShardName: func(obj interface{}) ([]string, error) {
+			if workspace, ok := obj.(*tenancyv1alpha1.Workspace); ok {
+				return []string{workspace.Spec.TargetShard}, nil
+			}
+			return []string{}, nil
+		},
+		unschedulable: func(obj interface{}) ([]string, error) {
+			if workspace, ok := obj.(*tenancyv1alpha1.Workspace); ok {
+				if IsWorkspaceUnschedulable(workspace) {
+					return []string{"true"}, nil
+				}
+			}
+			return []string{}, nil
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add indexer for Workspace: %w", err)
+	}
+
+	workspaceShardInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueAddedShard(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueDeletedShard(obj) },
+	})
+
+	return c, nil
+}
+
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	kcpClient        kcpclient.ClusterInterface
+	workspaceIndexer cache.Indexer
+	workspaceLister  tenancylister.WorkspaceLister
+
+	workspaceShardIndexer cache.Indexer
+	workspaceShardLister  tenancylister.WorkspaceShardLister
+}
+
+func (c *Controller) enqueue(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	klog.Infof("queueing workspace %q", key)
+	c.queue.AddRateLimited(key)
+}
+
+func (c *Controller) enqueueAddedShard(obj interface{}) {
+	shard, ok := obj.(*tenancyv1alpha1.WorkspaceShard)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("got %T when handling deleted WorkspaceShard", obj))
+		return
+	}
+	klog.Infof("handling added shard %q", shard.Name)
+	workspaces, err := c.workspaceIndexer.ByIndex(unschedulable, "true")
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	for _, workspace := range workspaces {
+		key, err := cache.MetaNamespaceKeyFunc(workspace)
+		if err != nil {
+			runtime.HandleError(err)
+			return
+		}
+		klog.Infof("queuing unschedulable workspace %q", key)
+		c.queue.AddRateLimited(key)
```

Comment:
Usually you call AddRateLimited when requeuing after an error, and Add() in an enqueue func.